saleor/discount/utils/voucher.py

```diff
@@ -525,12 +604,61 @@ def calculate_line_discount_amount_from_voucher(
     return discount_amount
 
 
+def calculate_order_line_discount_amount_from_denormalized_voucher(
+    line_info: "EditableOrderLineInfo", total_price: Money
+) -> Money:
+    """Calculate discount amount for line-level vouchers with denormalized values."""
+    voucher_info = line_info.voucher_denormalized_info
+    if not voucher_info:
+        return zero_money(total_price.currency)
+
+    quantity = line_info.line.quantity
+    currency = total_price.currency
+    if not voucher_info.apply_once_per_order:
+        if voucher_info.discount_value_type == DiscountValueType.PERCENTAGE:
+            discounted_total_price = apply_discount_to_value(
+                value=voucher_info.discount_value,
+                value_type=voucher_info.discount_value_type,
+                currency=currency,
+                price_to_discount=total_price,
+            )
+            voucher_discount_amount = max(
+                total_price - discounted_total_price, zero_money(total_price.currency)
+            )
+            discount_amount = min(voucher_discount_amount, total_price)
+        else:
+            unit_price = total_price / quantity
```

Comment:
Please also add a comment why the FIXED discount is applied on `unit_price` and why the `PERCENTAGE` on total price.